### ISSUE TITLE
Update Manifest to v3

### ISSUE THREAD

**Issue by phil-lgr** [rkodey/the-great-er-discarder-er#16]
![image](https://github.com/rkodey/the-great-er-discarder-er/assets/9644867/03a6c1e0-59f2-4f04-9d00-722f153b0c41)

**Comment by phil-lgr** [comment]
eventPagejs need to be rewritten without localStorage since it's not available in a service worker

![image](https://github.com/rkodey/the-great-er-discarder-er/assets/9644867/d9cae240-ef60-4482-b2fe-ff0c46784515)

**Comment by phil-lgr** [comment]
So @rkodey I took an hour to clone, install some types to help with extension development, upgrade to manifest v3, add webpack to compile

![image](https://github.com/rkodey/the-great-er-discarder-er/assets/9644867/2006652f-92c2-403a-ab24-4ab15eca6d40)

I am able to have a working output for the extension

Now for the source code:
- eventPagejs need to be rewritten without localStorage since it's not available in a service worker (I guess use the async chrome.storage https://developer.chrome.com/docs/extensions/reference/api/storage)
- db.js should probably be bundled from the dep in npm
- storage.js, tabsStates seem to work ok by replacing window to self (window global is not available in a service worker)

Let me know if you are interested in my contribution and if you have any thoughts on the point above

**Comment by rkodey** [comment]
Hello @phil-lgr - thanks for the reminder about manifest v3.

Since I originally took this over from the original abandoned extension to make a few tweaks... I never really dove deep into the inner workings, and I wonder if moving to the service worker might break in non-obvious ways.  If you have some commits I'd be happy to take a look and see what we can merge together.

In the mean time, I'll do a bit of the same adjustments you probably already have done, just to re-familiarize after not touching this extension for a few years!!

**Comment by vapier** [comment]
I ported the extension to MV3 and seems to be working for me.  I've started sending PRs, but more to test the waters to see if things actually get merged.  Most of the changes can be done in a MV2-compatible manner to make it easier to validate against current code before flipping the MV3 switch.

**Comment by CrisOrlandoBR** [comment]
Hey, any one could fork this and post a forked ? I am not a extension dev
Tks

**Comment by vapier** [comment]
my fork is unlisted at https://chromewebstore.google.com/detail/iibdpegbhfiibkfkdiecokimcakcgknc.  it's using MV3 and all that fun jazz.  but i'd really prefer to get changes merged back and not have people use mine :).

**Comment by CrisOrlandoBR** [comment]
Tks

**Comment by CrisOrlandoBR** [comment]
The other extensions for tab discarding are bugged and slow, this really works fine, tks

**Comment by rkodey** [comment]
Thanks for the work, @vapier !  I'm going to review the PRs today, and  also check the diffs against your fork.  Let's see if we can get an update!  I'm still amazed this extension remains useful after all these years...  I guess it's time for me to catch up on feature requests too!

**Comment by rkodey** [comment]
@vapier I actually don't see your MV3 changes.  I pulled the source from your unlisted extension above, but want to make sure you're ok if I merge your work.  At first glance, I like your approach.  I much prefer to make this manifest conversion as minimal as possible.  I'll look at converting to TypeScript separately.

**Comment by vapier** [comment]
I haven't sent PRs for everything, only some initial simple stuff. if you plan on merging, I can post the rest.

**Comment by rkodey** [comment]
Yea, I finally found some time this weekend, and it seems moving to V3 was a lot simpler than I initially assumed.  Thanks again for showing the way!

I think I might try to avoid some of the pure async refactoring, just to try to keep the V3 conversion truly as minimal as possible.  I'm not sure though - I'll run some tests tomorrow.

**Comment by vapier** [comment]
sent you more PRs.  they're designed to work with MV2 now so you can verify them, and they work when switching to MV3 so it avoids large one-off commits.

**Comment by rkodey** [comment]
Wow -- that was one of the easiest PR code reviews I've ever done!  Thanks!
Now, for testing...

**Comment by vapier** [comment]
I have one more commit locally to actually flip from MV2 to MV3, with minimum incompatible changes. I can't easily post it on its own since it depends on everything else. I've pushed the state to my main branch if you want to double check it. 

I also polished the profiler code, but that seems low priority, so I'll send that out once MV3 is done.

Hopefully the PR flood wasn't onerous. I find it easier for people to merge simpler things while debating more complicated changes so we don't end up with everything merges or nothing merges situation.

**Comment by rkodey** [comment]
PRs were awesome and easy!

Testing manifest v2, the Options did not seem to be saving to storage properly...  but I suppose we don't care about v2.  We'll go immediately to v3, so that's where I'm focused on testing.  I think I have the MV3 already running, but I'll grab your main branch and verify we don't have any material difference.

**Comment by rkodey** [comment]
Dev branch is here, if you're interested.
https://github.com/rkodey/the-great-er-discarder-er/tree/16-update-manifest-to-v3

Looks like you only missed 1-line in your PRs!  😉 
I'll tweak it on my side.

**Comment by rkodey** [comment]
Found a bug in the async messaging for the Save button in options.  No biggie, although it was hard/fun to unwind what was going on.  I think it's resolved.

I think we have a release candidate.
https://github.com/rkodey/the-great-er-discarder-er/tree/16-update-manifest-to-v3

Trying a couple more tests before merging into main.
I'll probably finish that up tomorrow.

**Comment by rkodey** [comment]
Well, I used the github recommended command line way of merging the PRs  ... and as a result it seems github doesn't see your contributions.  The commits are all there on the branch for this issue.

I've merged the branch into main, and will get the Chrome Web Store update later today.

**Comment by rkodey** [comment]
App submitted for review.  Now, we wait.  Closing this issue as fixed.

**Comment by vapier** [comment]
thx bud

**Comment by rkodey** [comment]
And we're already approved on the Chrome Web Store, and available to the public.  Nice!
Thanks again, to everyone on the thread!!
